drivers/adin2111/bm_adin2111.c

```diff
@@ -174,78 +212,219 @@ static void tx_complete(void *device_param, uint32_t event,
   free_tx_buffer(buffer_description);
 }
 
+#include "bcmp.h"
+#include "l2.h"
+#include "lwip/api.h"
+#include "lwip/debug.h"
+#include "lwip/ip.h"
+#include "lwip/nd6.h"
+#include "lwip/pbuf.h"
+#include "lwip/prot/ip6.h"
+#include "lwip/stats.h"
+#include "lwip/udp.h"
+#include "netif/ethernet.h"
+typedef struct {
+  struct eth_hdr eth_hdr;
+  struct ip6_hdr ip6_hdr;
+  uint8_t payload[0];
+} net_header_t;
+
+/*!
+  Add egress port to IP address and update UDP checksum
+
+  \param buff buffer with frame
+  \param port port in which frame is going out of
+  \return none
+*/
+static void add_egress_port(uint8_t *buff, uint8_t port) {
```

Comment:
@matt001k I believe that adding the egress port should be the packet management responsibility of L2, and that the driver should simply send the buffer it's given.

Reply:
I agree with you, it also is dependent on the number of ports that are available on the device and I was placing the logic that was previously in the `eth_adin2111.c` file here as is.
My plan is to try and move this up to L2, but this was step 1 in getting it to work properly.

Reply:
Got it. L2 should have access to the NetworkDevice now, which has a `num_ports` trait method in case that helps.

Reply:
Should be updated as of last push!